File: public/code/memecoin__tiered_exit.py
"""
Talons Strategies - Tiered Take Profit Exit (Memecoin)

STRATEJI:
- Memecoin'lerin hızlı pump-dump profiline uygun kademeli kar alma sistemi.
- Bir ENTRY stratejisi degil, POZISYON YONETIMI / EXIT katmanidir.
- +50%  -> %30 sat (anaparayi kurtar)
- +150% -> %30 sat
- +300% -> %30 sat
- Kalan %10 -> moonbag, trailing stop ile takip
- Trailing: ATH'den %35 dusus = full exit
- Hard SL: giris fiyatindan %25 dusus = full exit

INPUT (off-chain pozisyon durumu):
{
    "symbol": "TOKEN",
    "entry_price": float,
    "current_price": float,
    "position_pct_remaining": float,   # kalan pozisyon yuzdesi (0-100)
    "tier1_hit": bool,                  # +50% tetiklendi mi
    "tier2_hit": bool,                  # +150% tetiklendi mi
    "tier3_hit": bool,                  # +300% tetiklendi mi
    "ath_price": float,                 # pozisyon suresince gorulen ATH
}
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", ".."))
from common import make_signal


def check_signal(data: dict, cfg: dict = None) -> dict:
    cfg = cfg or {}
    tp1_mult = cfg.get("tp1_mult", 1.5)    # +50%
    tp2_mult = cfg.get("tp2_mult", 2.5)    # +150%
    tp3_mult = cfg.get("tp3_mult", 4.0)    # +300%
    sell1    = cfg.get("sell1_pct", 30.0)
    sell2    = cfg.get("sell2_pct", 30.0)
    sell3    = cfg.get("sell3_pct", 30.0)
    hard_sl  = cfg.get("hard_sl_pct", 25.0)
    trail_dd = cfg.get("trailing_dd_pct", 35.0)

    symbol  = data.get("symbol", "?")
    entry   = float(data.get("entry_price", 0))
    price   = float(data.get("current_price", 0))
    remain  = float(data.get("position_pct_remaining", 100.0))
    t1      = bool(data.get("tier1_hit", False))
    t2      = bool(data.get("tier2_hit", False))
    t3      = bool(data.get("tier3_hit", False))
    ath     = max(float(data.get("ath_price", entry)), price)

    if entry <= 0 or price <= 0:
        return make_signal("neutral", price, 0, [], 0, "Invalid entry/price")
    if remain <= 0:
        return make_signal("neutral", price, 0, [], 0, "Position already closed")

    pct_change = ((price - entry) / entry) * 100

    # Hard stop loss
    if pct_change <= -hard_sl:
        return make_signal(
            "short", price, 0, [], 1.0,
            f"HARD SL: {pct_change:.1f}% (full exit)",
            metadata={"symbol": symbol, "action": "full_exit", "sell_pct": remain},
        )

    # Trailing stop (sadece kar bolgesinde aktif)
    if pct_change > 30:
        dd_from_ath = ((ath - price) / ath) * 100
        if dd_from_ath >= trail_dd:
            return make_signal(
                "short", price, 0, [], 0.95,
                f"TRAILING STOP: {dd_from_ath:.1f}% drawdown from ATH (full exit)",
                metadata={"symbol": symbol, "action": "full_exit", "sell_pct": remain},
            )

    # Kademeli TP'ler
    if not t3 and price >= entry * tp3_mult:
        return make_signal(
            "short", price, 0, [], 0.9,
            f"TP3 +300%: sell {sell3}%",
            metadata={"symbol": symbol, "action": "partial_sell", "tier": 3, "sell_pct": sell3},
        )
    if not t2 and price >= entry * tp2_mult:
        return make_signal(
            "short", price, 0, [], 0.9,
            f"TP2 +150%: sell {sell2}%",
            metadata={"symbol": symbol, "action": "partial_sell", "tier": 2, "sell_pct": sell2},
        )
    if not t1 and price >= entry * tp1_mult:
        return make_signal(
            "short", price, 0, [], 0.9,
            f"TP1 +50%: sell {sell1}% (kurtar anapara)",
            metadata={"symbol": symbol, "action": "partial_sell", "tier": 1, "sell_pct": sell1},
        )

    return make_signal(
        "neutral", price, 0, [], 0,
        f"HOLD: {pct_change:+.1f}% (no tier hit)",
        metadata={"symbol": symbol, "action": "hold", "ath_price": ath, "pct_change": pct_change},
    )

File: public/code/memecoin__tiered_exit.ts
// Talons Strategies - Tiered Take Profit Exit
import { makeSignal, Signal } from "../../common";

export interface TieredExitData {
  symbol: string;
  entry_price: number;
  current_price: number;
  position_pct_remaining: number;
  tier1_hit: boolean;
  tier2_hit: boolean;
  tier3_hit: boolean;
}

export interface TieredExitConfig {
  tier1_gain_pct?: number;
  tier2_gain_pct?: number;
  tier3_gain_pct?: number;
  initial_sl_pct?: number;
  breakeven_trigger_pct?: number;
}

export function checkSignal(data: TieredExitData, cfg: TieredExitConfig = {}): Signal {
  const t1p = cfg.tier1_gain_pct ?? 50;
  const t2p = cfg.tier2_gain_pct ?? 150;
  const t3p = cfg.tier3_gain_pct ?? 300;
  const slP = cfg.initial_sl_pct ?? 30;
  const beT = cfg.breakeven_trigger_pct ?? 50;

  const { symbol, entry_price: entry, current_price: price,
          position_pct_remaining: rem,
          tier1_hit: t1, tier2_hit: t2, tier3_hit: t3 } = data;

  if (entry === 0 || rem <= 0)
    return makeSignal("neutral", price, 0, [], 0, "No position");

  const gain = ((price - entry) / entry) * 100;
  let sl = entry * (1 - slP/100);
  if (gain >= beT) sl = entry;

  if (gain <= -slP)
    return makeSignal("short", price, 0, [], 1.0,
                      `STOP LOSS (${gain.toFixed(1)}%)`,
                      { action: "exit_full", symbol });

  if (gain >= t3p && !t3 && rem > 0)
    return makeSignal("short", price, 0, [], 0.95,
                      `TP3 hit (${gain.toFixed(1)}%) - exit remaining`,
                      { action: "exit_remaining", tier: 3, symbol });
  if (gain >= t2p && !t2)
    return makeSignal("short", price, 0, [], 0.9,
                      `TP2 hit (${gain.toFixed(1)}%) - exit 33%`,
                      { action: "exit_33", tier: 2, symbol });
  if (gain >= t1p && !t1)
    return makeSignal("short", price, 0, [], 0.85,
                      `TP1 hit (${gain.toFixed(1)}%) - exit 33%`,
                      { action: "exit_33", tier: 1, symbol });

  return makeSignal("neutral", price, sl,
                    [entry*(1+t1p/100), entry*(1+t2p/100), entry*(1+t3p/100)],
                    0.5,
                    `Holding ${(rem*100).toFixed(0)}%, P&L ${gain.toFixed(1)}%`,
                    { symbol, gain_pct: gain });
}
